### PR TITLE
Add ability to reply to comments

### PR DESCRIPTION
**Pull request by IshaniSanturkar**
In this PR, I:
- added a feature that allows you to reply to existing comments, making trees of replies
The page now looks like this: Screenshot 2020-06-04 at 1 57 15 PM

In future PRs, I will work on optimizing the query to limit the number of comments displayed on the page since this is currently handled in the client. I will also add some means of recognizing the author of a reply to a comment, this is currently anonymous.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 1]
```diff
@@ -15,15 +15,18 @@
 package com.google.sps.servlets;
 
 import com.google.auto.value.AutoValue;
+// import java.util.ArrayList;
```
Comment (fenghaolw) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: revert?

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 2]
```diff
@@ -15,15 +15,18 @@
...
-    return new AutoValue_UserComment(name, email, comment, timestamp);
+  static UserComment create(String name, String email, String comment, long timestamp, long id, long parentId) {
+    return new AutoValue_UserComment(name, email, comment, timestamp, id, parentId);
   }
 
   abstract String name();
   abstract String email();
   abstract String comment();
   abstract long timestamp();
+  abstract long id();
```
Comment (fenghaolw) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java:64]: I think it is important to document what do these fields mean, how are they supposed to be used, and are there implicit restrictions (for example, is id always positive integer? is id required? is parentId optional? what are the default values if not 0? does negative integer have special meanings? you use -1 as a default value in other classes)

Without documentations, people need to go through the codebase to answer all of these questions. As you add more and more members to the value class, it becomes less and less clear to the readers.

File: portfolio/src/main/java/com/google/sps/servlets/ReplyServlet.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      long parentId = Long.parseLong(getFieldFromJsonObject(jsonObject, "parentid", "-1"));
+      addToDatastore(userName, userEmail, userDate, userComment, parentId);
+    }
+  }
+
+  /*
+   * Extracts the value of fieldName attribute from jsonObject if present
+   * and returns defaultValue if it is not or the value is empty
+   */
+  private String getFieldFromJsonObject(JsonObject jsonObject, String fieldName, String defaultValue) {
```
Comment (fenghaolw) [inline portfolio/src/main/java/com/google/sps/servlets/ReplyServlet.java]: [Optional] Some of these methods could be shared between ReplyServlet and DataServlet. Consider refactoring this in a followup PR

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -129,24 +129,73 @@ function loadComments() {
...
+    const commentTree = locateChildren(comments);
+    let numDisplayed = 0;
+    for (commentId in commentTree) {
+      if (numDisplayed == maxcomments) {
+        break;
+      }
+      let comment = commentTree[commentId];
+      if (comment["parentId"] === 0) {
+        numDisplayed++;
+        commentList.appendChild(constructReplyTree(comment, commentTree, 40));
```
Comment (fenghaolw) [inline portfolio/src/main/webapp/script.js:145]: why is this 40 instead of 20?

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -129,24 +129,73 @@ function loadComments() {
...
+      replyTree.appendChild(subTree);
+    }
+    replyTree.style.marginLeft = `${margin}px`;
+    thisReply.appendChild(replyTree);
+    return thisReply;
+  }
+}
+
 // Creates a list element with the given comment text and metadata (name, timestamp etc.)
 function createListElement(comment) {
```
Comment (fenghaolw) [inline portfolio/src/main/webapp/script.js:198]: [Optional] You might notice that it is tedious to do DOM manipulations in JS :) which is why most modern frontend frameworks try to help with this.

Nothing actionable here, but if you want you can play with some frontend frameworks (Angular, React, Vue, ...) to see how they make these code much easier to write and maintain.

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (IshaniSanturkar) [reply portfolio/src/main/webapp/script.js:145]: It represents the margin for replies to top-level comments. The margin for top-level comments is assigned to be 20 below and the margin increases by 20px for each level.

File: portfolio/src/main/java/com/google/sps/servlets/ReplyServlet.java (hunk 3 above)
Comment (IshaniSanturkar) [reply portfolio/src/main/java/com/google/sps/servlets/ReplyServlet.java]: Would it be okay to make the methods `public static` in one of the classes?

**Review by fenghaolw — APPROVED** [review]

**Pull request merged**